crates/simulacrum/src/checkpoint_builder.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use sui_types::{
+    base_types::VerifiedExecutionData,
+    effects::{TransactionEffects, TransactionEffectsAPI},
+    gas::GasCostSummary,
+    messages_checkpoint::{
+        CheckpointContents, CheckpointSummary, EndOfEpochData, VerifiedCheckpoint,
+    },
+    transaction::VerifiedTransaction,
+};
+
+use super::CommitteeWithKeys;
+
+pub struct CheckpointBuilder {
+    previous_checkpoint: VerifiedCheckpoint,
+    transactions: Vec<VerifiedExecutionData>,
+    epoch_rolling_gas_cost_summary: GasCostSummary,
+    epoch: u64,
+}
+
+impl CheckpointBuilder {
+    pub fn new(previous_checkpoint: VerifiedCheckpoint) -> Self {
+        let epoch_rolling_gas_cost_summary =
+            previous_checkpoint.epoch_rolling_gas_cost_summary.clone();
+        let epoch = previous_checkpoint.epoch;
+
+        Self {
+            previous_checkpoint,
+            transactions: Vec::new(),
+            epoch_rolling_gas_cost_summary,
+            epoch,
+        }
+    }
+
+    pub fn epoch_rolling_gas_cost_summary(&self) -> &GasCostSummary {
+        &self.epoch_rolling_gas_cost_summary
+    }
+
+    pub fn push_transaction(
+        &mut self,
+        transaction: VerifiedTransaction,
+        effects: TransactionEffects,
+    ) {
+        let gas_cost_summary = effects.gas_cost_summary();
+        self.epoch_rolling_gas_cost_summary.computation_cost += gas_cost_summary.computation_cost;
+        self.epoch_rolling_gas_cost_summary.storage_cost += gas_cost_summary.storage_cost;
+        self.epoch_rolling_gas_cost_summary.storage_rebate += gas_cost_summary.storage_rebate;
+        self.epoch_rolling_gas_cost_summary
+            .non_refundable_storage_fee += gas_cost_summary.non_refundable_storage_fee;
+
+        self.transactions
+            .push(VerifiedExecutionData::new(transaction, effects))
+    }
+
+    /// Builds a checkpoint using internally buffered transactions.
+    pub fn build(
+        &mut self,
+        committee: &CommitteeWithKeys<'_>,
+        timestamp_ms: u64,
+    ) -> (VerifiedCheckpoint, CheckpointContents) {
+        let contents =
+            CheckpointContents::new_with_causally_ordered_execution_data(self.transactions.iter());
+
+        let summary = CheckpointSummary {
+            epoch: self.epoch,
```

Comment:
I feel this and `build_end_of_epoch` could be refactored/combined into one fn.

Reply:
I'll see what i can do